cpp/src/groupby/groupby.cu

```diff
@@ -108,6 +114,89 @@ void verify_valid_requests(std::vector<aggregation_request> const& requests) {
+ * column of the same type as the input. Therefore it is suggested to convert integer column types
+ * (especially low-precision integers) to `FLOAT32` or `FLOAT64` before doing a rolling `MEAN`.
+ *
+ * @param[in] input_col The input column
+ * @param[in] group_offsets A column of indexes, indicating partition/grouping boundaries.
+ * @param[in] preceding_window The static rolling window size in the backward direction.
+ * @param[in] following_window The static rolling window size in the forward direction.
+ * @param[in] min_periods Minimum number of observations in window required to have a value,
+ *                        otherwise element `i` is null.
+ * @param[in] op The rolling window aggregation type (SUM, MAX, MIN, etc.)
+ *
+ * @returns   A nullable output column containing the rolling window results
+ **/
+
+std::unique_ptr<column> rolling_window(column_view const& input,
+                                       rmm::device_vector<cudf::size_type> const& group_offsets,
+                                       size_type preceding_window,
+                                       size_type following_window,
+                                       size_type min_periods,
+                                       std::unique_ptr<aggregation> const& aggr,
+                                       rmm::mr::device_memory_resource* mr)
+{
+  if (group_offsets.empty()) {
+    // Empty group_offsets list. Treat `input` as a single group. i.e. Ignore grouping.
+    return rolling_window(input, preceding_window, following_window, min_periods, aggr, mr);
+  }
+
+  // `group_offsets` are interpreted in adjacent pairs, each pair representing the offsets
+  // of the first, and one past the last elements in a group.
+  //
+  // If `group_offsets` is not empty, it must contain at least two offsets:
+  //   a. 0, indicating the first element in `input`
+  //   b. input.size(), indicating one past the last element in `input`.
+  //
+  // Thus, for an input of 1000 rows,
+  //   0. [] indicates a single group, spanning the entire column.
+  //   1  [10] is invalid.
+  //   2. [0, 1000] indicates a single group, spanning the entire column (thus, equivalent to no groups.)
+  //   3. [0, 500, 1000] indicates two equal-sized groups: [0,500), and [500,1000).
+
+  CUDF_EXPECTS(group_offsets.size() >= 2 && group_offsets[0] == 0 
+               && group_offsets[group_offsets.size()-1] == input.size(),
+               "Must have at least one group.");
+
+  auto offsets_begin = group_offsets.begin(); // Required, since __device__ lambdas cannot capture by ref,
+  auto offsets_end   = group_offsets.end();   //   or capture local variables without listing them.
+
+  auto preceding_calculator = [offsets_begin, offsets_end, preceding_window] __device__ (size_type idx) {
+    // `upper_bound()` cannot return `offsets_end`, since it is capped with `input.size()`.
+    auto group_end = thrust::upper_bound(thrust::seq, offsets_begin, offsets_end, idx);
```

Comment:
Why is `upper_bound` being used here at all? That's a lot more expensive than the algorithm I suggested before with doing:
```c++
gid = group_id[i]
start = offsets[gid]
stop = offsets[gid+1]
intra_group_index = i - start
preceding[i] = (intra_group_index < preceding) ? intra_group_index : preceding;
```

Reply:
I might've misread your suggested algorithm, at the time. Let me try that again.